File: MinecraftManhunt/src/main/java/com/yoonicode/minecraftmanhuntplus/respawn_inventory/InventoryGenerator.java
package com.yoonicode.minecraftmanhuntplus.respawn_inventory;

import com.yoonicode.minecraftmanhuntplus.PluginMain;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Random;

public class InventoryGenerator {
    PluginMain main;
    private int level;
    private double modifier;
    private Random random = new Random();
    // Fields

    // 0-1
    private ArrayList<ItemStack> respawnOne = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.WOODEN_AXE),
            new ItemStack(Material.WOODEN_PICKAXE),
            new ItemStack(Material.WOODEN_SHOVEL),
            new ItemStack(Material.WOODEN_SWORD),
            new ItemStack(Material.COAL,5),
            new ItemStack(Material.BREAD, 5),
            new ItemStack(Material.TORCH,5)
    ));

    // 1-2
    private ArrayList<ItemStack> respawnTwo = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.STONE_AXE),
            new ItemStack(Material.STONE_PICKAXE),
            new ItemStack(Material.STONE_SHOVEL),
            new ItemStack(Material.STONE_SWORD),
            new ItemStack(Material.IRON_HELMET)
    ));

    // 2-3
    private ArrayList<ItemStack> respawnThree = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.STONE_AXE),
            new ItemStack(Material.IRON_PICKAXE),
            new ItemStack(Material.IRON_SHOVEL),
            new ItemStack(Material.STONE_SWORD),
            new ItemStack(Material.GOLDEN_CHESTPLATE),
            new ItemStack(Material.GOLDEN_LEGGINGS),
            new ItemStack(Material.BREAD,10),
            new ItemStack(Material.TORCH,10)
    ));

    // 3-4
    private ArrayList<ItemStack> respawnFour = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.IRON_AXE),
            new ItemStack(Material.IRON_PICKAXE),
            new ItemStack(Material.STONE_SHOVEL),
            new ItemStack(Material.STONE_SWORD),
            new ItemStack(Material.IRON_HELMET),
            new ItemStack(Material.IRON_BOOTS),
            new ItemStack(Material.FURNACE),
            new ItemStack(Material.COAL, 3),
            new ItemStack(Material.OAK_LOG,10)
    ));

    // 4-5
    private ArrayList<ItemStack> respawnFive = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.IRON_AXE),
            new ItemStack(Material.IRON_PICKAXE),
            new ItemStack(Material.STONE_SHOVEL),
            new ItemStack(Material.STONE_SWORD),
            new ItemStack(Material.IRON_HELMET),
            new ItemStack(Material.IRON_BOOTS),
            new ItemStack(Material.GOLDEN_CHESTPLATE),
            new ItemStack(Material.GOLDEN_LEGGINGS),
            new ItemStack(Material.FURNACE),
            new ItemStack(Material.COAL,3),
            new ItemStack(Material.OAK_LOG,2),
            new ItemStack(Material.COOKED_BEEF,5)
    ));

    // 5-6
    private ArrayList<ItemStack> respawnSix = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.IRON_AXE),
            new ItemStack(Material.IRON_PICKAXE),
            new ItemStack(Material.STONE_SHOVEL),
            new ItemStack(Material.STONE_SWORD),
            new ItemStack(Material.IRON_HELMET),
            new ItemStack(Material.IRON_BOOTS),
            new ItemStack(Material.GOLDEN_CHESTPLATE),
            new ItemStack(Material.IRON_LEGGINGS),
            new ItemStack(Material.FURNACE),
            new ItemStack(Material.COAL,3),
            new ItemStack(Material.OAK_LOG,2),
            new ItemStack(Material.COOKED_BEEF,10)
    ));

    // 6-7
    private ArrayList<ItemStack> respawnSeven = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.IRON_AXE),
            new ItemStack(Material.IRON_PICKAXE),
            new ItemStack(Material.STONE_SHOVEL),
            new ItemStack(Material.STONE_SWORD),
            new ItemStack(Material.IRON_HELMET),
            new ItemStack(Material.IRON_BOOTS),
            new ItemStack(Material.IRON_CHESTPLATE),
            new ItemStack(Material.IRON_LEGGINGS),
            new ItemStack(Material.FURNACE),
            new ItemStack(Material.COAL,3),
            new ItemStack(Material.OAK_LOG,2),
            new ItemStack(Material.COOKED_BEEF,5)
    ));

    // 7-8
    private ArrayList<ItemStack> respawnEight = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.IRON_AXE),
            new ItemStack(Material.IRON_PICKAXE),
            new ItemStack(Material.STONE_SHOVEL),
            new ItemStack(Material.STONE_SWORD),
            new ItemStack(Material.IRON_HELMET),
            new ItemStack(Material.IRON_BOOTS),
            new ItemStack(Material.IRON_CHESTPLATE),
            new ItemStack(Material.IRON_LEGGINGS),
            new ItemStack(Material.FURNACE),
            new ItemStack(Material.COAL,3),
            new ItemStack(Material.OAK_LOG,2),
            new ItemStack(Material.COOKED_BEEF,10),
            new ItemStack(Material.SHIELD)
    ));

    // 8-9
    private ArrayList<ItemStack> respawnNine = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.IRON_AXE),
            new ItemStack(Material.IRON_PICKAXE),
            new ItemStack(Material.STONE_SHOVEL),
            new ItemStack(Material.STONE_SWORD),
            new ItemStack(Material.IRON_HELMET),
            new ItemStack(Material.IRON_BOOTS),
            new ItemStack(Material.IRON_CHESTPLATE),
            new ItemStack(Material.IRON_LEGGINGS),
            new ItemStack(Material.FURNACE),
            new ItemStack(Material.COAL,3),
            new ItemStack(Material.OAK_LOG,2),
            new ItemStack(Material.COOKED_BEEF,15),
            new ItemStack(Material.SHIELD),
            new ItemStack(Material.BOW),
            new ItemStack(Material.ARROW,10)
    ));

    private ArrayList<ItemStack> tierOne = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.IRON_ORE, random.nextInt(5)+1),
            new ItemStack(Material.BREAD, random.nextInt(5)+5)
    ));

    private ArrayList<ItemStack> tierTwo = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.IRON_HELMET),
            new ItemStack(Material.IRON_BOOTS),
            new ItemStack(Material.IRON_ORE, random.nextInt(5)+1),
            new ItemStack(Material.BREAD, random.nextInt(5)+5),
            new ItemStack(Material.COAL, random.nextInt(5)+5),
            new ItemStack(Material.OAK_LOG, random.nextInt(3)+5)
            ));

    private ArrayList<ItemStack> tierThree = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.BREAD, random.nextInt(5)+1),
            new ItemStack(Material.COAL, random.nextInt(10)+1),
            new ItemStack(Material.OAK_LOG, random.nextInt(8)+5),
            new ItemStack(Material.IRON_ORE, random.nextInt(5)+1),
            new ItemStack(Material.TORCH, 32)
            ));

    private ArrayList<ItemStack> tierFour = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.BREAD, random.nextInt(5)+1),
            new ItemStack(Material.COAL, random.nextInt(10)+1),
            new ItemStack(Material.OAK_LOG, random.nextInt(5)+1),
            new ItemStack(Material.IRON_ORE, random.nextInt(5)+1),
            new ItemStack(Material.ENDER_PEARL)
    ));

    private ArrayList<ItemStack> tierFive = new ArrayList<>(Arrays.asList(
            new ItemStack(Material.ENDER_PEARL),
            new ItemStack(Material.DIAMOND, random.nextInt(2))
    ));


    private ArrayList<ArrayList<ItemStack>> respawnLevel = new ArrayList<>(Arrays.asList(
            respawnOne,
            respawnTwo,
            respawnThree,
            respawnFour,
            respawnFive,
            respawnSix,
            respawnSeven,
            respawnEight,
            respawnNine
    ));

    private ArrayList<ArrayList<ItemStack>> effectTier = new ArrayList<>(Arrays.asList(
            tierOne,
            tierTwo,
            tierThree,
            tierFour,
            tierFive
    ));


    public InventoryGenerator(PluginMain main) {
        this.main = main;
    }


    public ArrayList<ItemStack> generateItemStack() {
        calculateScore(main.getGameState());
        ArrayList<ItemStack> mainItems = new ArrayList<>();
        mainItems.addAll(respawnLevel.get(level));
        boolean addItems = Math.random() < this.modifier;
        ArrayList<ItemStack> extraItems = new ArrayList<>();
        if (addItems){
            ArrayList<ItemStack> activeTier;
            if (level == 0){
                activeTier = effectTier.get(0);
            }
            else {
                activeTier = effectTier.get(level-1);
            }
            for (ItemStack itemStack : activeTier) {
                if (Math.random() < modifier) {
                    extraItems.add(itemStack);
                }
            }
        }
        mainItems.addAll(extraItems);
        return mainItems;
    }

    private void calculateScore(double score) {
        this.level = (int) score;
        String scoreAsString = Double.toString(score);
        String decimalString = scoreAsString.substring(scoreAsString.indexOf("."));
        this.modifier = getDecimal(decimalString);
    }

    private double getDecimal(String s){
        String fullDecimal = "0" + s;
        return Double.parseDouble(fullDecimal);
    }

}

File: MinecraftManhunt/src/main/java/com/yoonicode/minecraftmanhuntplus/respawn_inventory/InventoryGenerator.kt
package com.yoonicode.minecraftmanhuntplus.respawn_inventory

import com.yoonicode.minecraftmanhuntplus.PluginMain
import org.bukkit.inventory.ItemStack
import org.bukkit.Material
import java.util.*

/**
 * TODO
 * Generates the inventory that a respawned player gets.
 *
 * @property main the plugin main being used
 */
class InventoryGenerator(var main: PluginMain) {
    private var level = 0
    private var modifier = 0.0
    private val random = Random()

    // Fields
    // 0-1
    private val respawnOne = listOf(
            ItemStack(Material.WOODEN_AXE),
            ItemStack(Material.WOODEN_PICKAXE),
            ItemStack(Material.WOODEN_SHOVEL),
            ItemStack(Material.WOODEN_SWORD),
            ItemStack(Material.COAL, 5),
            ItemStack(Material.BREAD, 5),
            ItemStack(Material.TORCH, 5)
    )

    // 1-2
    private val respawnTwo = listOf(
            ItemStack(Material.STONE_AXE),
            ItemStack(Material.STONE_PICKAXE),
            ItemStack(Material.STONE_SHOVEL),
            ItemStack(Material.STONE_SWORD),
            ItemStack(Material.IRON_HELMET)
    )

    // 2-3
    private val respawnThree = listOf(
            ItemStack(Material.STONE_AXE),
            ItemStack(Material.IRON_PICKAXE),
            ItemStack(Material.IRON_SHOVEL),
            ItemStack(Material.STONE_SWORD),
            ItemStack(Material.GOLDEN_CHESTPLATE),
            ItemStack(Material.GOLDEN_LEGGINGS),
            ItemStack(Material.BREAD, 10),
            ItemStack(Material.TORCH, 10)
    )

    // 3-4
    private val respawnFour = listOf(
            ItemStack(Material.IRON_AXE),
            ItemStack(Material.IRON_PICKAXE),
            ItemStack(Material.STONE_SHOVEL),
            ItemStack(Material.STONE_SWORD),
            ItemStack(Material.IRON_HELMET),
            ItemStack(Material.IRON_BOOTS),
            ItemStack(Material.FURNACE),
            ItemStack(Material.COAL, 3),
            ItemStack(Material.OAK_LOG, 10)
    )

    // 4-5
    private val respawnFive = listOf(
            ItemStack(Material.IRON_AXE),
            ItemStack(Material.IRON_PICKAXE),
            ItemStack(Material.STONE_SHOVEL),
            ItemStack(Material.STONE_SWORD),
            ItemStack(Material.IRON_HELMET),
            ItemStack(Material.IRON_BOOTS),
            ItemStack(Material.GOLDEN_CHESTPLATE),
            ItemStack(Material.GOLDEN_LEGGINGS),
            ItemStack(Material.FURNACE),
            ItemStack(Material.COAL, 3),
            ItemStack(Material.OAK_LOG, 2),
            ItemStack(Material.COOKED_BEEF, 5)
    )

    // 5-6
    private val respawnSix = listOf(
            ItemStack(Material.IRON_AXE),
            ItemStack(Material.IRON_PICKAXE),
            ItemStack(Material.STONE_SHOVEL),
            ItemStack(Material.STONE_SWORD),
            ItemStack(Material.IRON_HELMET),
            ItemStack(Material.IRON_BOOTS),
            ItemStack(Material.GOLDEN_CHESTPLATE),
            ItemStack(Material.IRON_LEGGINGS),
            ItemStack(Material.FURNACE),
            ItemStack(Material.COAL, 3),
            ItemStack(Material.OAK_LOG, 2),
            ItemStack(Material.COOKED_BEEF, 10)
    )

    // 6-7
    private val respawnSeven = listOf(
            ItemStack(Material.IRON_AXE),
            ItemStack(Material.IRON_PICKAXE),
            ItemStack(Material.STONE_SHOVEL),
            ItemStack(Material.STONE_SWORD),
            ItemStack(Material.IRON_HELMET),
            ItemStack(Material.IRON_BOOTS),
            ItemStack(Material.IRON_CHESTPLATE),
            ItemStack(Material.IRON_LEGGINGS),
            ItemStack(Material.FURNACE),
            ItemStack(Material.COAL, 3),
            ItemStack(Material.OAK_LOG, 2),
            ItemStack(Material.COOKED_BEEF, 5)
    )

    // 7-8
    private val respawnEight = listOf(
            ItemStack(Material.IRON_AXE),
            ItemStack(Material.IRON_PICKAXE),
            ItemStack(Material.STONE_SHOVEL),
            ItemStack(Material.STONE_SWORD),
            ItemStack(Material.IRON_HELMET),
            ItemStack(Material.IRON_BOOTS),
            ItemStack(Material.IRON_CHESTPLATE),
            ItemStack(Material.IRON_LEGGINGS),
            ItemStack(Material.FURNACE),
            ItemStack(Material.COAL, 3),
            ItemStack(Material.OAK_LOG, 2),
            ItemStack(Material.COOKED_BEEF, 10),
            ItemStack(Material.SHIELD)
    )

    // 8-9
    private val respawnNine = listOf(
            ItemStack(Material.IRON_AXE),
            ItemStack(Material.IRON_PICKAXE),
            ItemStack(Material.STONE_SHOVEL),
            ItemStack(Material.STONE_SWORD),
            ItemStack(Material.IRON_HELMET),
            ItemStack(Material.IRON_BOOTS),
            ItemStack(Material.IRON_CHESTPLATE),
            ItemStack(Material.IRON_LEGGINGS),
            ItemStack(Material.FURNACE),
            ItemStack(Material.COAL, 3),
            ItemStack(Material.OAK_LOG, 2),
            ItemStack(Material.COOKED_BEEF, 15),
            ItemStack(Material.SHIELD),
            ItemStack(Material.BOW),
            ItemStack(Material.ARROW, 10)
    )

    private val tierOne = listOf(
            ItemStack(Material.IRON_ORE, random.nextInt(5) + 1),
            ItemStack(Material.BREAD, random.nextInt(5) + 5)
    )

    private val tierTwo = listOf(
            ItemStack(Material.IRON_HELMET),
            ItemStack(Material.IRON_BOOTS),
            ItemStack(Material.IRON_ORE, random.nextInt(5) + 1),
            ItemStack(Material.BREAD, random.nextInt(5) + 5),
            ItemStack(Material.COAL, random.nextInt(5) + 5),
            ItemStack(Material.OAK_LOG, random.nextInt(3) + 5)
    )

    private val tierThree = listOf(
            ItemStack(Material.BREAD, random.nextInt(5) + 1),
            ItemStack(Material.COAL, random.nextInt(10) + 1),
            ItemStack(Material.OAK_LOG, random.nextInt(8) + 5),
            ItemStack(Material.IRON_ORE, random.nextInt(5) + 1),
            ItemStack(Material.TORCH, 32)
    )

    private val tierFour = listOf(
            ItemStack(Material.BREAD, random.nextInt(5) + 1),
            ItemStack(Material.COAL, random.nextInt(10) + 1),
            ItemStack(Material.OAK_LOG, random.nextInt(5) + 1),
            ItemStack(Material.IRON_ORE, random.nextInt(5) + 1),
            ItemStack(Material.ENDER_PEARL)
    )

    private val tierFive = listOf(
            ItemStack(Material.ENDER_PEARL),
            ItemStack(Material.DIAMOND, random.nextInt(2))
    )

    private val respawnLevel = listOf(
            respawnOne,
            respawnTwo,
            respawnThree,
            respawnFour,
            respawnFive,
            respawnSix,
            respawnSeven,
            respawnEight,
            respawnNine
    )

    private val effectTier = listOf(
            tierOne,
            tierTwo,
            tierThree,
            tierFour,
            tierFive
    )

    /**
     * Generates a list of ItemStacks to be added to the inventory
     *
     * Generates standard items based on the game score, and a chance at
     * extra items if the game score's based on the game score's decimals
     *
     * @return a list of ItemStacks
     */
    fun generateInventory(): List<ItemStack> {
        calculateScore(main.gameState)
        val mainItems = mutableListOf<ItemStack>()
        mainItems.addAll(respawnLevel[level])
        val addItems = Math.random() < modifier
        val extraItems = mutableListOf<ItemStack>()
        if (addItems) {
            val activeTier: List<ItemStack> = if (level == 0) {
                effectTier[0]
            } else {
                effectTier[level - 1]
            }
            for (itemStack in activeTier) {
                if (Math.random() < modifier) {
                    extraItems.add(itemStack)
                }
            }
        }
        mainItems.addAll(extraItems)
        return mainItems
    }

    /**
     * Calculates the score
     *
     * @param score
     */
    private fun calculateScore(score: Double) {
        level = score.toInt()
        val scoreAsString = java.lang.Double.toString(score)
        val decimalString = scoreAsString.substring(scoreAsString.indexOf("."))
        modifier = getDecimal(decimalString)
    }

    /**
     * Gets the decimal value in the game score
     *
     * @param s game score
     * @return the decimal value
     */
    private fun getDecimal(s: String): Double {
        val fullDecimal = "0$s"
        return fullDecimal.toDouble()
    }
}